src/crypto/chacha20_vec.ipp

```diff
@@ -0,0 +1,342 @@
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/chacha20_vec.h>
+
+#include <bit>
+#include <cassert>
+#include <cstring>
+#include <limits>
+
+#if defined(ENABLE_CHACHA20_VEC)
+
+#if defined(CHACHA20_VEC_DISABLE_STATES_16) && \
+    defined(CHACHA20_VEC_DISABLE_STATES_8) && \
+    defined(CHACHA20_VEC_DISABLE_STATES_6) && \
+    defined(CHACHA20_VEC_DISABLE_STATES_4) && \
+    defined(CHACHA20_VEC_DISABLE_STATES_2)
+#define CHACHA20_VEC_ALL_MULTI_STATES_DISABLED
+#endif
+
+
+#if !defined(CHACHA20_VEC_ALL_MULTI_STATES_DISABLED)
+
+#if defined(__has_attribute)
+#  if __has_attribute(always_inline)
+#    define ALWAYS_INLINE __attribute__ ((always_inline)) inline
+#  endif
+#endif
+
+#if !defined(ALWAYS_INLINE)
+#  define ALWAYS_INLINE inline
+#endif
+
+
+namespace {
+
+using vec256 = uint32_t __attribute__((__vector_size__(32)));
+
+/** Endian-conversion for big-endian */
+ALWAYS_INLINE void vec_byteswap(vec256& vec)
+{
+    if constexpr (std::endian::native == std::endian::big)
+    {
+        vec256 ret;
+        ret[0] = __builtin_bswap32(vec[0]);
+        ret[1] = __builtin_bswap32(vec[1]);
+        ret[2] = __builtin_bswap32(vec[2]);
+        ret[3] = __builtin_bswap32(vec[3]);
+        ret[4] = __builtin_bswap32(vec[4]);
+        ret[5] = __builtin_bswap32(vec[5]);
+        ret[6] = __builtin_bswap32(vec[6]);
+        ret[7] = __builtin_bswap32(vec[7]);
+        vec = ret;
+    }
+}
+
+/** Left-rotate vector */
+template <size_t BITS>
+ALWAYS_INLINE void vec_rotl(vec256& vec)
+{
+    vec = (vec << BITS) | (vec >> (32 - BITS));
+}
+
+/** Store a vector in all array elements */
+template <size_t I, size_t ITER = 0>
+ALWAYS_INLINE void arr_set_vec256(std::array<vec256, I>& arr, const vec256& vec)
+{
+    std::get<ITER>(arr) = vec;
+    if constexpr(ITER + 1 < I ) arr_set_vec256<I, ITER + 1>(arr, vec);
+}
+
+/** Add a vector to all array elements */
+template <size_t I, size_t ITER = 0>
+ALWAYS_INLINE void arr_add_vec256(std::array<vec256, I>& arr, const vec256& vec)
+{
+    std::get<ITER>(arr) += vec;
+    if constexpr(ITER + 1 < I ) arr_add_vec256<I, ITER + 1>(arr, vec);
+}
+
+/** Add corresponding vectors in arr1 to arr0 */
+template <size_t I, size_t ITER = 0>
+ALWAYS_INLINE void arr_add_arr(std::array<vec256, I>& arr0, const std::array<vec256, I>& arr1)
+{
+    std::get<ITER>(arr0) += std::get<ITER>(arr1);
+    if constexpr(ITER + 1 < I ) arr_add_arr<I, ITER + 1>(arr0, arr1);
+}
+
+/** Perform add/xor/rotate for the round function */
+template <size_t BITS, size_t I, size_t ITER = 0>
+ALWAYS_INLINE void arr_add_xor_rot(std::array<vec256, I>& arr0, const std::array<vec256, I>& arr1, std::array<vec256, I>& arr2)
+{
+    vec256& x = std::get<ITER>(arr0);
+    const vec256& y = std::get<ITER>(arr1);
+    vec256& z = std::get<ITER>(arr2);
+
+    x += y;
+    z ^= x;
+    vec_rotl<BITS>(z);
+
+    if constexpr(ITER + 1 < I ) arr_add_xor_rot<BITS, I, ITER + 1>(arr0, arr1, arr2);
+}
```

Comment:
Hmmm, it seems to me we're doing heavy recursive template instantiationshere to unroll loops.
My understanding (and experience with the mentioned Obfuscation PR) is that modern compilers are good at unrolling fixed-bound loops.
Can you please try if this also works and results in the same speedup?

```suggestion
template <size_t BITS, size_t I>
ALWAYS_INLINE void arr_add_xor_rot(std::array<vec256, I>& arr0, const std::array<vec256, I>& arr1, std::array<vec256, I>& arr2)
{
    for (size_t i{0}; i < I; ++i) {
        arr0[i] += arr1[i];
        arr2[i] ^= arr0[i];
        vec_rotl<BITS>(arr2[i]);
    }
}
```